### PR TITLE
Issue #6479: please support preceding, preceding-sibling, following a…

### PR DESCRIPTION
**Pull request by timurt**
https://github.com/checkstyle/checkstyle/issues/6479

### CODE REVIEW

**Comment by rnveach** [comment]
@timurt pitest is failing
> `+ElementNode.java.html:<td class='covered'><pre><span  class='survived'>        if (getFollowingSibling() == null) {</span></pre></td></tr>`
`+ElementNode.java.html:<td class='covered'><pre><span  class='survived'>        if (getPrecedingSibling() == null) {</span></pre></td></tr>`

**Comment by timurt** [comment]
@rnveach 

> +ElementNode.java.html:<td class='covered'><pre><span class='survived'> if (getFollowingSibling() == null) {</span></pre></td></tr>
> +ElementNode.java.html:<td class='covered'><pre><span class='survived'> if (getPrecedingSibling() == null) {</span></pre></td></tr>

fixed

**Comment by romani** [comment]
@pbludov , please do first review

**Comment by pbludov** [comment]
> @pbludov , please do first review

The whole design is non-obvious. Now it is a double linked list stored in the array list. I may be wrong, but since the children list is immutable (appendable only), it is sufficient to store only the node index here. Then the next node will be `getParent().children[idx++]`. The iterator of followings is the iterator of the children of parent from `idx` to `size` and so on.

@romani @rnveach what are you think of it?

**Comment by romani** [comment]
https://www.saxonica.com/html/documentation/javadoc/net/sf/saxon/om/NodeInfo.html#getParent--

So AbstractNode do know parent.
Looks like reasonable to operate by indexes.

@timurt , please defend your solution, if no objections with indexes, please implement.

**Comment by romani** [comment]
Please rebase on latest master to fix Travis build if it fail.

**Comment by romani** [comment]
@timurt , please rebase to resolve conflict.

**Comment by romani** [comment]
@timurt , please let me know when you come back from summer break to continue this PR.

**Comment by timurt** [comment]
@romani @pbludov rewrote implementation, please check

**Comment by romani** [comment]
while we are looking, please fix TC violation (move `++` increment to next line, as separate statement)

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 1]
```diff
@@ -64,13 +70,15 @@
...
+     * @param position position of the current node among its siblings
      */
-    public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst) {
+    public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst, int position) {
         super(root.getTreeInfo());
         this.parent = parent;
         this.root = root;
         this.detailAst = detailAst;
         text = TokenUtil.getTokenName(detailAst.getType());
+        positionAmongSiblings = position;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: Can we do `parent.getChildren().size() - 1` here ?

It is better to name field as `index` `indexAmongSiblings`, as position word does not imply strict sequence.

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: Can we get index from parent node ? `parent.getChildren().size() - 1`

It is better to name field as `index...` to imply that it is strict sequence

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 2]
```diff
@@ -269,6 +294,68 @@ public DetailAST getUnderlyingNode() {
...
+        return result;
+    }
+
+    /**
+     * Returns following siblings of the current node.
+     * @return siblings
+     */
+    private List<AbstractNode> getFollowingSiblings() {
+        final List<AbstractNode> siblings = new ArrayList<>();
+        for (int i = positionAmongSiblings + 1; i < parent.getChildren().size(); i++) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: Is it the same as
```
final List<AbstractNode> siblings = parent.getChildren();
return siblings.subList(positionAmongSiblings, siblings.size());
```
? The list returned by parent.getChildren() is immutable, so it is safe to use sublist here.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 3]
```diff
@@ -269,6 +294,68 @@ public DetailAST getUnderlyingNode() {
...
+        }
+        return siblings;
+    }
+
+    /**
+     * Returns preceding siblings of the current node.
+     * @return siblings
+     */
+    private List<AbstractNode> getPrecedingSiblings() {
+        final List<AbstractNode> siblings = new ArrayList<>();
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: same

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 2 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 3 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 1 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 4]
```diff
@@ -55,6 +57,9 @@
     /** Represents text of the DetailAST. */
     private final String text;
 
+    /** Represents text of the DetailAST. */
+    private final int positionAmongSiblings;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: this was not done.

from my comment above:
>It is better to name field as index indexAmongSiblings, as position word does not imply strict sequence.

Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: sorry, done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java [hunk 5]
```diff
@@ -358,5 +358,4 @@ public final void copy(Receiver receiver, int index, Location location) {
     private static UnsupportedOperationException throwUnsupportedOperationException() {
         return new UnsupportedOperationException("Operation is not supported");
     }
-
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java]: Please undo this change. A blank line before type's closing brace is on purpose.

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 6]
```diff
@@ -419,7 +419,7 @@ public void testQueryRootNotImplementedAxis() throws Exception {
 
     @Test
     public void testQueryElementNotImplementedAxis() throws Exception {
-        final String xpath = "/CLASS_DEF//following-sibling::METHOD_DEF";
+        final String xpath = "/CLASS_DEF//namespace::*";
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: should we mention in doc that https://www.w3schools.com/xml/xpath_axes.asp "namespace" axis will result in exception (not supported as there is no namespaces) ? Or it is kind of obvious ?

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 7]
```diff
@@ -786,6 +786,169 @@ public void testFourthCaseGroup() throws Exception {
...
+        final DetailAST expectedSemiNode2 = expectedVariableDefNode2.getNextSibling();
+        final DetailAST[] expected = {expectedSemiNode2, expectedSemiNode1,
+            expectedVariableDefNode2, expectedVariableDefNode1};
+        assertArrayEquals("Result nodes differ from expected", expected, actual);
+    }
+
+    @Test
+    public void testQueryElementPrecedingSiblingVariableDef() throws Exception {
+        final String xpath = "//VARIABLE_DEF[./IDENT[@text='array']]/preceding-sibling::"
+                + "VARIABLE_DEF";
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: please make it:
```
        final String xpath = "//VARIABLE_DEF[./IDENT[@text='array']]"
 + "/preceding-sibling::VARIABLE_DEF";
```
to cut by xpath element, it will be much easy to read.

Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java (hunk 5 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 6 above)
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: we can mention all supported axes in `config_filters`

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: it would be awesome, please do in any form you like it to be (link to all possible will be good to have). If some not supported, it would be good to put brief details on why (some reason, or link to issue tracker if it is our todo list).

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 8]
```diff
@@ -786,6 +786,169 @@ public void testFourthCaseGroup() throws Exception {
...
+        final String xpath = "//METHOD_DEF/following-sibling::*";
+        final RootNode rootNode = getRootNode("InputXpathMapperAst.java");
+        final DetailAST[] actual = convertToArray(getXpathItems(xpath, rootNode));
+        final DetailAST expectedMethodDefNode = getSiblingByType(rootNode.getUnderlyingNode(),
+                TokenTypes.CLASS_DEF)
+                .findFirstToken(TokenTypes.OBJBLOCK)
+                .findFirstToken(TokenTypes.METHOD_DEF)
+                .getNextSibling();
+        final DetailAST[] expected = {expectedMethodDefNode,
+                expectedMethodDefNode.getNextSibling()};
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: it might be personal, but it took me some time to verify this, `expectedMethodDefNode.getNextSibling()` is very abstract. So human can not verify this test, without execution of it or deep dive in Input AST structure.
Does it make sense to extend test with `assertThat("...", RCURLY,  expectedMethodDefNode.getNextSibling().getType())`;
So users we see what is expected. Yes , flollowing `assertArrayEquals` will do this, but with extra assert test expected result could be verified by human.

**Comment by romani** [comment]
@timurt , ping.

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 9]
```diff
@@ -786,6 +786,169 @@ public void testFourthCaseGroup() throws Exception {
...
+        assertArrayEquals("Result nodes differ from expected", expected, actual);
+    }
+
+    @Test
+    public void testQueryElementPrecedingLiteralPublic() throws Exception {
+        final String xpath = "//LITERAL_CLASS/preceding::LITERAL_PUBLIC";
+        final RootNode rootNode = getRootNode("InputXpathMapperPreceding.java");
+        final DetailAST[] actual = convertToArray(getXpathItems(xpath, rootNode));
+        final DetailAST expectedLiteralPublicNode = getSiblingByType(rootNode.getUnderlyingNode(),
+                TokenTypes.CLASS_DEF).getFirstChild().getFirstChild();
```
[the remark this answers is not in the document]
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: The main query is `//LITERAL_CLASS/preceding`, selects all nodes that appear before `LITERAL_CLASS`

`//LITERAL_CLASS/preceding::LITERAL_PUBLIC`, select all `LITERAL_PUBLIC` nodes that appear before `LITERAL_CLASS`

[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: Please educate me on " Selects all nodes that appear before the current node in the document"
in tree `LITERAL_PUBLIC` is as child, so it is after/below "LITERAL_CLASS". https://www.merriam-webster.com/dictionary/preceding 
Where I am missing a nuance in terminology ?

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 6 above)
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: just added list of supported axes, only unsupported is `namespace` axis

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 10]
```diff
@@ -786,6 +786,169 @@ public void testFourthCaseGroup() throws Exception {
...
+                .findFirstToken(TokenTypes.SLIST)
+                .findFirstToken(TokenTypes.VARIABLE_DEF);
+        final DetailAST[] expected = {expectedVariableDefNode};
+        assertArrayEquals("Result nodes differ from expected", expected, actual);
+    }
+
+    @Test
+    public void testQueryElementPrecedingOne() throws Exception {
+        final String xpath = "//LITERAL_CLASS/preceding::*";
+        final RootNode rootNode = getRootNode("InputXpathMapperPreceding.java");
```
[the remark this answers is not in the document]
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java [hunk 11]
```diff
@@ -786,6 +786,169 @@ public void testFourthCaseGroup() throws Exception {
...
+            expectedMethodDefNode};
+        assertArrayEquals("Result nodes differ from expected", expected, actual);
+    }
+
+    @Test
+    public void testQueryElementNoFollowing() throws Exception {
+        final String xpath = "//CLASS_DEF/following::*";
+        final RootNode rootNode = getRootNode("InputXpathMapperAst.java");
+        final DetailAST[] actual = convertToArray(getXpathItems(xpath, rootNode));
+        assertEquals("Invalid number of nodes", 0, actual.length);
```
[the remark this answers is not in the document]
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 8 above)
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: > in tree LITERAL_PUBLIC is as child, so it is after/below "LITERAL_CLASS".

No. Actually `LITERAL_PUBLIC` appears before `LITERAL_CLASS` in document order, it is like DFS traverse

```
CLASS_DEF -> CLASS_DEF [3:0]
|--MODIFIERS -> MODIFIERS [3:0]
|   `--LITERAL_PUBLIC -> public [3:0]  
|--LITERAL_CLASS -> class [3:7]         
```

[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: My bad, misread LITERAL_CLASS and CLASS_DEF

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: Please update all other adder 'assertXxxxx'

**Comment by rnveach** [comment]
Something is wrong with CI, travis didn't run and a few items have weird failures. I recommend repushing.

**Comment by romani** [comment]
@timurt , ping.

**Comment by timurt** [comment]
@romani @rnveach rebased

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: still not done.

@timurt , please move all unrelated to issue refactoring in separate commit of this PR.
just do:
```
git reset HEAD~1
git add -p
```
keep assertThat in funtional fix only you added to cover issue fix.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java]: still not done.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 2 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: verified, done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: verfied, done.

**Comment by romani** [comment]
@timurt , please review 2 items that are still not done.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java (hunk 5 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java]: now it is done)

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (timurt) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: please check, I tried to move assertThat changes to separate commit

[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java]: good, thanks a lot.

advice for future: such minor commits it is better to put as first commit because chance that it require additional fix is very small, but functional fix might require additional change. Amending non top commit is not trivial in git.

In your case .... if some changes are requested in first commit, you can use https://roman-ivanov.blogspot.com/2019/05/shell-function-to-put-changes-in.html

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java [hunk 12]
```diff
@@ -337,6 +337,47 @@
  * <a href="https://github.com/checkstyle/checkstyle/issues/4830">issue 4830</a>.
  * </p>
  * <p>
+ * Currently, filter supports the following xpath axes:
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java]: Just FYI for others: https://www.saxonica.com/html/documentation/expressions/axissteps.html
